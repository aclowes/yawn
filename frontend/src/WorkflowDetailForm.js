import React from 'react';
import {
  Checkbox, FormGroup, FormControl, ControlLabel,
  Button, Alert, Pagination
} from 'react-bootstrap';

import API from './API'

/* Helper to extract form values from a given object */
function formValues(object) {
  let parameters;
  // convert the parameters between 'bash' and 'dict' syntax
  if (typeof object.parameters === "string") {
    parameters = {};
    object.parameters.split(/\r\n|\r|\n/g).forEach((line) => {
      const parts = line.split('=');
      if (parts.length === 2) parameters[parts[0]] = parts[1];
    })
  } else {
    parameters = Object.keys(object.parameters).map((key) => {
      return `${key}=${object.parameters[key]}`
    });
    parameters = (parameters || []).join('\n');  // empty object returns null
  }
  return {
    schedule: object.schedule,
    schedule_active: object.schedule_active,
    parameters: parameters,
  }
}

export default class WorkflowDetailForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      editable: false,
      error: null,
      versions: [],
      ...formValues(this.props.workflow)
    };
  }

  componentDidMount() {
    API.get(`/api/names/${this.props.workflow.name_id}/`, (payload, error) => {
      this.setState({versions: payload.versions, error})
    });
  }

  componentWillReceiveProps(nextProps) {
    // the version changed...
    if (nextProps.workflow !== this.props.workflow) {
      this.setState({...formValues(nextProps.workflow)});
    }
  }

  selectVersion = (eventKey) => {
    const version = this.state.versions[eventKey - 1];
    this.props.router.push(`/workflows/${version}`);
  };

  toggleEditable = () => {
    this.setState({editable: !this.state.editable});
  };

  toggleCheckbox = (event) => {
    this.setState({[event.target.value]: event.target.checked});
  };

  handleChange = (event) => {
    this.setState({[event.target.id]: event.target.value});
  };

  handleSubmit = (event) => {
    const update = formValues(this.state);
    API.patch(`/api/workflows/${this.props.workflow.id}/`, update, (payload, error) => {
      this.setState({...formValues(payload), error, editable: false});
    });
    event.preventDefault();
  };

  render() {
    if (this.state.editable) {
      return (
        <form onSubmit={this.handleSubmit}>
          {this.state.error && <Alert bsStyle="danger">{this.state.error}</Alert>}
          <Checkbox value="schedule_active" onChange={this.toggleCheckbox}
                    checked={this.state.schedule_active}>Schedule Active</Checkbox>
          <FormGroup controlId="schedule">
            <ControlLabel>Schedule</ControlLabel>
            <FormControl type="text" placeholder="<minute> <hour> <weekday>"
                         onChange={this.handleChange} value={this.state.schedule}/>
          </FormGroup>
          <FormGroup controlId="parameters">
            <ControlLabel>Parameters</ControlLabel>
            <FormControl componentClass="textarea" placeholder="VARIABLE=value (one per line)"
                         onChange={this.handleChange} value={this.state.parameters}/>
          </FormGroup>
          <Button type="submit" bsStyle="primary">Save</Button>
          <Button onClick={this.toggleEditable}>Cancel</Button>
        </form>
      );
    } else {
      return (
        <div>
          <dl className="dl-horizontal">
            <dt>Name</dt>
            <dd>{this.props.workflow.name}</dd>
            <dt>Version</dt>
            <dd>{}
              <Pagination
                ellipsis
                bsSize="small"
                items={this.state.versions.length}
                maxButtons={10}
                activePage={this.props.workflow.version}
                onSelect={this.selectVersion}/>
            </dd>
            <dt>Schedule Active</dt>
            <dd>{this.state.schedule_active ? 'True' : 'False'}</dd>
            <dt>Schedule</dt>
            <dd>{this.state.schedule}</dd>
            <dt>Next Run</dt>
            <dd>{this.props.workflow.next_run}</dd>
            <dt>Parameters</dt>
            <dd>
              <pre>{this.state.parameters}</pre>
            </dd>
          </dl>
          <Button onClick={this.toggleEditable}>Edit</Button>
        </div>
      )
    }
  }
}