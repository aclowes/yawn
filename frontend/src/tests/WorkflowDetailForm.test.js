import React from 'react';
import ReactTestRenderer from 'react-test-renderer';

import WorkflowDetailForm from '../WorkflowDetailForm';
import {mockAPI} from './mocks'

it('WorkflowDetailForm editable', () => {
  mockAPI([{versions: [1]}]);
  const workflow = {
      "id": 1,
      "name": "Simple Workflow Example",
      "name_id": 1,
      "version": 1,
      "schedule_active": false,
      "schedule": null,
      "next_run": null,
      "parameters": {"MY_OBJECT_ID": "1", "SOME_SETTING": "false"}
    };
  const component = ReactTestRenderer.create(<WorkflowDetailForm workflow={workflow}/>);
  component.getInstance().toggleEditable();
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
