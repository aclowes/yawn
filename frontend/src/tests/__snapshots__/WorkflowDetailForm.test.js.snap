// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkflowDetailForm editable 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="checkbox"
    style={undefined}
  >
    <label>
      <input
        checked={false}
        disabled={false}
        onChange={[Function]}
        type="checkbox"
        value="schedule_active"
      />
      Schedule Active
    </label>
  </div>
  <div
    className="form-group"
  >
    <label
      className="control-label"
      htmlFor="schedule"
    >
      Schedule
    </label>
    <input
      className="form-control"
      id="schedule"
      onChange={[Function]}
      placeholder="<minute> <hour> <weekday>"
      type="text"
      value={null}
    />
  </div>
  <div
    className="form-group"
  >
    <label
      className="control-label"
      htmlFor="parameters"
    >
      Parameters
    </label>
    <textarea
      className="form-control"
      id="parameters"
      onChange={[Function]}
      placeholder="VARIABLE=value (one per line)"
      type={undefined}
      value="MY_OBJECT_ID=1
    SOME_SETTING=false"
    />
  </div>
  <div
    className="btn-toolbar"
    role="toolbar"
  >
    <button
      className="btn btn-default"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Cancel
    </button>
    <button
      className="btn btn-primary"
      disabled={false}
      type="submit"
    >
      Save
    </button>
  </div>
</form>
`;
